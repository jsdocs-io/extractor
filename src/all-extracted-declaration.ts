import type {
  ExtractedClass,
  ExtractedClassConstructor,
  ExtractedClassMethod,
  ExtractedClassProperty,
} from "./extract-class";
import type { ExtractedEnum, ExtractedEnumMember } from "./extract-enum";
import type { ExtractedFunction } from "./extract-function";
import type {
  ExtractedInterface,
  ExtractedInterfaceCallSignature,
  ExtractedInterfaceConstructSignature,
  ExtractedInterfaceGetAccessor,
  ExtractedInterfaceIndexSignature,
  ExtractedInterfaceMethod,
  ExtractedInterfaceProperty,
  ExtractedInterfaceSetAccessor,
} from "./extract-interface";
import type { ExtractedNamespace } from "./extract-namespace";
import type { ExtractedTypeAlias } from "./extract-type-alias";
import type { ExtractedVariable } from "./extract-variable";

/**
`AllExtractedDeclaration` is the union of all possible declarations
that can be extracted, with some being found only in other declarations
(e.g., class method declarations are found only in a class declaration).
*/
export type AllExtractedDeclaration =
  | ExtractedVariable
  | ExtractedFunction
  | ExtractedClass
  | ExtractedClassConstructor
  | ExtractedClassProperty
  | ExtractedClassMethod
  | ExtractedInterface
  | ExtractedInterfaceProperty
  | ExtractedInterfaceMethod
  | ExtractedInterfaceConstructSignature
  | ExtractedInterfaceCallSignature
  | ExtractedInterfaceIndexSignature
  | ExtractedInterfaceGetAccessor
  | ExtractedInterfaceSetAccessor
  | ExtractedEnum
  | ExtractedEnumMember
  | ExtractedTypeAlias
  | ExtractedNamespace;

/**
`AllExtractedDeclarationKind` is the union of all discriminators
used to detect the kind of declaration.
*/
export type AllExtractedDeclarationKind = AllExtractedDeclaration["kind"];
