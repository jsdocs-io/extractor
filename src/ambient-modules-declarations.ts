import { ModuleDeclaration, Node, type Project } from "ts-morph";
import { isHidden } from "./is-hidden";
import { sourceFilePath } from "./source-file-path";

export type AmbientModulesDeclarationsReturn = {
  containerName: string;
  exportName: string;
  declaration: ModuleDeclaration;
}[];

export const ambientModulesDeclarations = (
  containerName: string,
  project: Project,
  pkgName?: string,
): AmbientModulesDeclarationsReturn => {
  const ambientModulesDeclarations = [];
  for (const symbol of project.getAmbientModules()) {
    for (const declaration of symbol.getDeclarations()) {
      if (isHidden(declaration) || !Node.isModuleDeclaration(declaration)) {
        continue;
      }
      if (pkgName && !sourceFilePath(declaration).startsWith(`/${pkgName}`)) {
        // Ignore ambient modules that are not from the analyzed package.
        continue;
      }
      const exportName = declaration.getName();
      // TODO:
      // Remove surrounding quotes and eventual spaces
      // const exportID = exportName
      //   .replace(/"|'/g, "")
      //   .replace(/\s/g, "_")
      //   .trim();
      ambientModulesDeclarations.push({
        containerName,
        exportName,
        declaration,
      });
    }
  }
  return ambientModulesDeclarations;
};
