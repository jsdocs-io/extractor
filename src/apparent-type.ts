import { TypeFormatFlags, ts, type Node } from "ts-morph";

export function apparentType(node: Node): string {
	// See:
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-667578386
	return node
		.getType()
		.getApparentType()
		.getText(
			node,
			ts.TypeFormatFlags.NoTruncation | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope,
		)
		.replace(/^Number$/, "number")
		.replace(/^Boolean$/, "boolean")
		.replace(/^String$/, "string");
}
