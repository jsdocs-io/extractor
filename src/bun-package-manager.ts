import { Effect } from "effect";
import { execa } from "execa";
import { InstallPackageError, PackageManager } from "./package-manager";

/** @internal */
export const bunPackageManager = (bunPath = "bun") =>
	PackageManager.of({
		installPackage: ({ pkg, cwd }) =>
			Effect.gen(function* (_) {
				// Run `bun add <pkg> --verbose`.
				// See https://bun.sh/docs/cli/add.
				const { stdout } = yield* _(
					Effect.tryPromise({
						try: () => execa(bunPath, ["add", pkg, "--verbose"], { cwd }),
						catch: (e) => new InstallPackageError({ cause: e }),
					}),
				);

				// With verbose output on, bun prints one line per installed package
				// (e.g., "foo@1.0.0"), including all installed dependencies.
				// These lines are between the two delimiting lines found here:
				// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
				const lines = stdout.split("\n");
				const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
				const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
				const installedPackages = lines.slice(beginHash + 1, endHash);
				return installedPackages;
			}),
	});
