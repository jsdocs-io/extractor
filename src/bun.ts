import { execa } from "execa";

/** `Bun` is a wrapper for the Bun package manager. */
export class Bun {
	#bunCmd: string;

	constructor(bunCmd = "bun") {
		this.#bunCmd = bunCmd.trim();
	}

	/**
	`add` installs the given package in the given directory and
	returns a list of all installed packages (e.g., `["foo@1.0.0"]`).

	@see {@link https://bun.sh/docs/cli/add}
	*/
	async add(pkg: string, cwd: string): Promise<string[]> {
		// Run `bun add <pkg> --verbose` in the `cwd`.
		const { stdout } = await execa(this.#bunCmd, ["add", pkg, "--verbose"], { cwd });

		// With verbose output on, bun prints one line per installed package
		// (e.g., "foo@1.0.0"), including all installed dependencies.
		// These lines are between the two delimiting lines found here:
		// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
		const lines = stdout.split("\n");
		const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
		const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
		const installedPackages = lines.slice(beginHash + 1, endHash);
		return installedPackages;
	}
}
