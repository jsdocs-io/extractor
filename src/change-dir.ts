import { Result } from "neverthrow";
import { OsError } from "./errors";

export const changeDir = Result.fromThrowable(
  process.chdir,
  (e) => new OsError("failed to change directory", { cause: e }),
);
