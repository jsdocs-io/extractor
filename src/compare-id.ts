export type WithId = {
  id: string;
};

export const compareId = (a: WithId, b: WithId): number =>
  a.id.localeCompare(b.id);
