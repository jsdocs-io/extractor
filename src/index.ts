export {
  extractDeclarations,
  type ExtractDeclarationsOptions,
  type ExtractedDeclaration,
  type ExtractedDeclarationKind,
} from "./extract-declarations";
