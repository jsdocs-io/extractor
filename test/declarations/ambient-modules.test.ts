import { dedent } from "ts-dedent";
import { ModuleKind, ModuleResolutionKind, Project, ScriptTarget } from "ts-morph";
import { expect, test } from "vitest";
import { extractDeclarations } from "../../src";

test("ambient modules", async () => {
	const project = new Project({
		useInMemoryFileSystem: true,
		compilerOptions: {
			lib: ["lib.esnext.full.d.ts"],
			target: ScriptTarget.ESNext,
			module: ModuleKind.ESNext,
			moduleResolution: ModuleResolutionKind.Bundler,
		},
	});
	const indexFile = project.createSourceFile(
		"index.d.ts",
		dedent`
    /** Module 'foo' */
    declare module "foo";

    /** Module 'bar' */
    declare module "bar" {
      /** Variable var1 */
      const var1: string;

      /** Variable var2 */
      export const var2: boolean;
    }

    /** Module 'foo bar' */
    declare module "foo bar" {
      export default function sum(a: number, b: number): number;
    }
    `,
	);
	expect(
		await extractDeclarations({
			containerName: "",
			container: indexFile,
			maxDepth: 5,
			project,
		}),
	).toMatchSnapshot();
});
