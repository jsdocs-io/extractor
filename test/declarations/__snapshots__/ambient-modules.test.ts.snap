// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ambient modules 1`] = `
[
  {
    "declarations": [
      {
        "docs": [
          "/** Variable var1 */",
        ],
        "file": "/index.d.ts",
        "id": "+namespace."bar".+variable.var1",
        "kind": "variable",
        "line": 7,
        "name": "var1",
        "signature": "const var1: string;",
      },
      {
        "docs": [
          "/** Variable var2 */",
        ],
        "file": "/index.d.ts",
        "id": "+namespace."bar".+variable.var2",
        "kind": "variable",
        "line": 10,
        "name": "var2",
        "signature": "const var2: boolean;",
      },
    ],
    "docs": [
      "/** Module 'bar' */",
    ],
    "file": "/index.d.ts",
    "id": "+namespace."bar"",
    "kind": "namespace",
    "line": 5,
    "name": ""bar"",
    "signature": "module "bar" {}",
  },
  {
    "declarations": [
      {
        "docs": [],
        "file": "/index.d.ts",
        "id": "+namespace."foo bar".+function.default",
        "kind": "function",
        "line": 15,
        "name": "default",
        "signature": "default: (a: number, b: number) => number;",
      },
    ],
    "docs": [
      "/** Module 'foo bar' */",
    ],
    "file": "/index.d.ts",
    "id": "+namespace."foo bar"",
    "kind": "namespace",
    "line": 14,
    "name": ""foo bar"",
    "signature": "module "foo bar" {}",
  },
  {
    "declarations": [],
    "docs": [
      "/** Module 'foo' */",
    ],
    "file": "/index.d.ts",
    "id": "+namespace."foo"",
    "kind": "namespace",
    "line": 2,
    "name": ""foo"",
    "signature": "module "foo" {}",
  },
]
`;
